pkg/k8s/service_cache.go

```diff
@@ -0,0 +1,384 @@
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/versioned"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+)
+
+// CacheAction is the type of action that was performed on the cache
+type CacheAction int
+
+const (
+	// UpdateService reflects that the service was updated or added
+	UpdateService CacheAction = iota
+
+	// DeleteService reflects that the service was deleted
+	DeleteService
+
+	// UpdateIngress reflects that the ingress was updated or added
+	UpdateIngress
+
+	// DeleteIngress reflects that the ingress was deleted
+	DeleteIngress
+)
+
+// String returns the cache action as a string
+func (c CacheAction) String() string {
+	switch c {
+	case UpdateService:
+		return "service-updated"
+	case DeleteService:
+		return "service-deleted"
+	case UpdateIngress:
+		return "ingress-updated"
+	case DeleteIngress:
+		return "ingress-deleted"
+	default:
+		return "unknown"
+	}
+}
+
+// ServiceEvent is emitted via the Events channel of ServiceCache and describes
+// the change that occurred in the cache
+type ServiceEvent struct {
+	// Action is the action that was performed in the cache
+	Action CacheAction
+
+	// ID is the identified of the service
+	ID ServiceID
+
+	// Service is the service structure
+	Service *Service
+
+	// K8sService is the corresponding Kubernetes service resource that
+	// triggered the action or nil if the action was not triggered by a
+	// Kubernetes service update
+	K8sService *v1.Service
+
+	// Endpoints is the endpoints structured correlated with the service
+	Endpoints *Endpoints
+
+	// K8sEndpoints is the corresponding Kubernetes endpoints resource that
+	// triggered the action or nil if the action was not triggered by a
+	// Kubernetes endpoints update
+	K8sEndpoints *v1.Endpoints
+}
+
+// ServiceCache is a list of services and ingresses correlated with the
+// matching endpoints. The Events member will receive events as services and
+// ingresses
+type ServiceCache struct {
+	mutex     lock.RWMutex
+	services  map[ServiceID]*Service
+	endpoints map[ServiceID]*Endpoints
+	ingresses map[ServiceID]*Service
+
+	Events chan ServiceEvent
+}
+
+// NewServiceCache returns a new ServiceCache
+func NewServiceCache() ServiceCache {
+	return ServiceCache{
+		services:  map[ServiceID]*Service{},
+		endpoints: map[ServiceID]*Endpoints{},
+		ingresses: map[ServiceID]*Service{},
+		Events:    make(chan ServiceEvent, 128),
+	}
+}
+
+// UpdateService parses a Kubernetes service and adds or updates it in the
+// ServiceCache. Returns the ServiceID unless the Kubernetes service could not
+// be parsed and a bool to indicate whether the service was changed in the
+// cache or not.
+func (s *ServiceCache) UpdateService(k8sSvc *v1.Service) ServiceID {
+	svcID, newService := ParseService(k8sSvc)
+	if newService == nil {
+		return svcID
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if oldService, ok := s.services[svcID]; ok {
+		if oldService.DeepEquals(newService) {
```

Comment:
Create either a GH issue of a function that knows how to perform the deep equal of v1.Service for the fields that actually matter to us.

Reply:
I don't understand. This is the same code as before just moved into a separate package.

Reply:
Discussed offline. Was mistaken for using `reflect.DeepEquals()`